evennia/utils/tests/test_funcparser.py

```diff
@@ -216,12 +238,48 @@ def test_parse(self, string, expected):
         # print(f"time: {(t1-t0)*1000} ms")
         self.assertEqual(expected, ret)
 
-    def test_parse_raise(self):
+    @parameterized.expand(
+        (
+            "Test malformed This is $dummy(a, b) and $bar(",
+            "Test $funcNotFound()",
+        )
+    )
+    def test_parse_raise_unparseable(self, unparseable):
         """
         Make sure error is raised if told to do so.
 
         """
-        string = "Test malformed This is $dummy(a, b) and $bar("
+        with self.assertRaises(funcparser.ParsingError):
+            self.parser.parse(unparseable, raise_errors=True)
+
+    @patch("evennia.utils.funcparser._MAX_NESTING", 2)
+    def test_parse_max_nesting(self):
+        """
+        Make sure it is an error if the max nesting value is reached.
+
+        TODO: Does this make sense? When it sees the first function, len(callstack)
+        is 0. It doesn't raise until the stack length is greater than the
+        _MAX_NESTING value, which means you can nest 4 values with a value of
+        2, as demonstrated by this test.
+        """
+        string = "$add(1, $add(1, $add(1, $add(1, $toint(42)))))"
+        ret = self.parser.parse(string)
+
+        # TODO: Does this return value actually make sense?
+        # It removed the spaces from the calls.
+        self.assertEqual("$add(1,$add(1,$add(1,$add(1,$toint(42)))))", ret)
```

Comment:
The entire `$(...)` will be replaced in-place by whatever is returned by that outermost `$()`. So whitespace inside arguments are skipped; that is why it compresses the failed output. I agree it would look better to keep the original whitespace when letting the parsing fail like this though.

Reply:
I think I noticed this whitespace collapsing when returning any function that had multiple arguments that failed parsing, with `raise_errors=False`, but I don't have test which demonstrates it.